### PR TITLE
Return Promise for set and remove

### PR DESCRIPTION
**Pull request by WyriHaximus**
This PR builds on #16 by resolving merge conflicts and adding more documentation in the cache interface plus adding documentation to the readme.

Supersedes / closes #16

### CODE REVIEW

File: src/ArrayCache.php [hunk 1]
```diff
@@ -20,10 +20,12 @@ public function get($key)
     public function set($key, $value)
     {
         $this->data[$key] = $value;
+        return new Promise\FulfilledPromise(true);
```
Comment (jsor) [inline src/ArrayCache.php]: I'd prefer `Promise\resolve(true)` here.

File: src/ArrayCache.php [hunk 2]
```diff
@@ -20,10 +20,12 @@ public function get($key)
...
     }
 
     public function remove($key)
     {
         unset($this->data[$key]);
+        return new Promise\FulfilledPromise(true);
```
Comment (jsor) [inline src/ArrayCache.php]: I'd prefer `Promise\resolve(true)` here.

File: src/CacheInterface.php [hunk 3]
```diff
@@ -16,19 +21,22 @@
...
 
     /**
-     * Store an item in the cache.
+     * Store an item in the cache, returns a promise which resolves to true on success or
+     * rejects with an exception on error.
      *
      * @param string $key
      * @param mixed $value
-     * @return void
+     * @return PromiseInterface<bool|Exception>
```
Comment (jsor) [inline src/CacheInterface.php]: As discussed (ot sure where atm), only fulfillment values should be included.

`PromiseInterface<bool>`

File: src/CacheInterface.php [hunk 4]
```diff
@@ -16,19 +21,22 @@
...
 
     /**
-     * Remove an item from the cache.
+     * Remove an item from the cache, returns a promise which resolves to true on success or
+     * rejects with an exception on error. When the $key isn't found in the cache it also
+     * resolves true.
      *
      * @param string $key
-     * @return void
+     * @return PromiseInterface<bool|Exception>
```
Comment (jsor) [inline src/CacheInterface.php]: `PromiseInterface<bool>` (see above)

**Comment by WyriHaximus** [comment]
@jsor addressed your points :+1:

File: src/CacheInterface.php [hunk 5]
```diff
@@ -4,6 +4,11 @@
 
 use React\Promise\PromiseInterface;
 
+/**
+ * All methods only reject with an exception when an error occurs on the underlying storage
+ * layer, for example the connection with Redis dropped and cannot be recovered, or the
+ * directory on the filesystem used to store cache has the wrong permissions of is full.
```
Comment (jsor) [inline src/CacheInterface.php]: _of_ is full -> _or_ is full

**Comment by WyriHaximus** [comment]
@jsor fixed :+1:

**Comment by jsor** [comment]
Good point, there was also a discussion here: https://github.com/reactphp/cache/pull/16/files#r104752102

**Comment by WyriHaximus** [comment]
@jsor @clue updated PR with the points discussed during our call

**Comment by jsor** [comment]
Regarding `set()` now returning `null`: That differs now from PSR-16. As @clued noted in https://github.com/reactphp/cache/pull/23#pullrequestreview-99473991:

> `@return bool True on success and false on failure.`

But since i've not attended the call, i'm probably missing something.

**Comment by WyriHaximus** [comment]
@jsor right so the thing @clue and I discussed is that when it always resolves true on success or rejects with an error we might as well resolve with null because true doesn't add any value.

**Comment by jsor** [comment]
The point is, that according to PSR-16, `set()` **never** throws, except for an invalid key. That means, our promise **only** rejects for an invalid key and resolves to `true`/`false` otherwise.

Refs:
* https://groups.google.com/d/msg/php-fig/kSj_yVbkwOw/zJwzYPaeAwAJ
* https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Simple/AbstractCache.php#L117-L123
* https://github.com/reactphp/cache/pull/16/files#r104756818

**Comment by WyriHaximus** [comment]
@jsor that makes sense so we should always resolve with `true` on success `false` on failure. That will mean the use won't get any exceptions but the cache implementation should handle/log those them self. Makes perfect sense to me. What do you think @clue ?

**Comment by clue** [comment]
We've discussed this only briefly in our last hangouts. My point is: I can see some valid arguments for both directions as rightfully pointed out by @jsor. I personally don't have a use case that requires any specific behavior, I only require consistent behavior.

For now, I would vote to keep this in line with PSR-16 and simply use a promise with a boolean resolution value and basically never reject.

I would rather get this feature out and we can still discuss this for a future version :shipit:

**Comment by WyriHaximus** [comment]
Ping @jsor && @clue updated the PR

File: src/CacheInterface.php [hunk 6]
```diff
@@ -4,31 +4,40 @@
...
+ * layer, for example the connection with Redis dropped and cannot be recovered, or the
+ * directory on the filesystem used to store cache has the wrong permissions or is full.
+ */
 interface CacheInterface
 {
     /**
      * Retrieve an item from the cache, resolves with its value on
-     * success or null when no item can be found.
+     * success or null when no item can be found. It will reject with an exception
+     * on error. (Note that a cache miss isn't an error.)
```
Comment (clue) [inline src/CacheInterface.php]: Rejecting would be the async equivalent of an `Exception` and I don't see this in PSR-16?

File: src/CacheInterface.php [hunk 7]
```diff
@@ -4,31 +4,40 @@
 
 use React\Promise\PromiseInterface;
 
+/**
+ * All methods only reject with an exception when an error occurs on the underlying storage
+ * layer, for example the connection with Redis dropped and cannot be recovered, or the
+ * directory on the filesystem used to store cache has the wrong permissions or is full.
+ */
```
Comment (clue) [inline src/CacheInterface.php]: Not sure if this is a leftover, but this seems to contradict the below definitions?

File: src/CacheInterface.php [hunk 8]
```diff
@@ -4,31 +4,40 @@
...
 
     /**
-     * Store an item in the cache.
+     * Store an item in the cache, returns a promise which resolves to true on success or
+     * false on error.
      *
      * @param string $key
      * @param mixed $value
-     * @return void
+     * @return PromiseInterface<bool>
```
Comment (clue) [inline src/CacheInterface.php]: I don't see any notation for this in any of the major projects, so I would suggest using the more verbose form `@return PromiseInterface Returns a promise which resolves to true on success of false on error` and possibly update the description above? (See also below)

File: README.md [hunk 9]
```diff
@@ -47,7 +47,8 @@ This example fetches the value of the key `foo` and passes it to the
 `var_dump` function. You can use any of the composition provided by
 [promises](https://github.com/reactphp/promise).
 
-If the key `foo` does not exist, the promise will be fulfilled with `null` as value.
+If the key `foo` does not exist, the promise will be fulfilled with `null` as value. On 
+any error it will reject with an exception.
```
Comment (clue) [inline README.md]: leftover?

**Comment by WyriHaximus** [comment]
@clue whoops updated

**Comment by WyriHaximus** [comment]
When @jsor also doesn't find any issues and approves I'll squash everything 👍

File: README.md [hunk 10]
```diff
@@ -56,8 +57,9 @@ $cache->set('foo', 'bar');
 ```
 
 This example eventually sets the value of the key `foo` to `bar`. If it
-already exists, it is overridden. No guarantees are made as to when the cache
-value is set. If the cache implementation has to go over the network to store
+already exists, it is overridden. To provide guarantees as to when the cache
+value is set a promise is returned. Which will fulfill with `true` on success or
```
Comment (jsor) [inline README.md]: This should either be one sentence ("...a promise is returned which will fulfill...") or better separated ("...a promise is returned. The promise will fulfill...").

**Comment by WyriHaximus** [comment]
@jsor done

**Comment by WyriHaximus** [comment]
Squashed the commits, the whole history without the squash is available here: https://github.com/WyriHaximus-labs/cache/tree/return_promises_history

**Pull request merged**